### DISCUSSION TITLE
GC runtime instance lifetimes

### DISCUSSION THREAD

**Discussion by CharlieTap** [WebAssembly/gc#567] [Q&A]
Struct and Array Instances are allocated in the Store, the stores reference to the instances prevents these objects from being GC'd until the store itself is deallocated (In my particular runtime impl). As an optimisation I store a [strong reference to the struct instance hidden in the execution value](https://github.com/CharlieTap/chasm/blob/a452ab9e37504517b43118f636a28f54e7bcb57c/executor/runtime/src/commonMain/kotlin/io/github/charlietap/chasm/executor/runtime/value/ReferenceValue.kt#L20) and make the reference to the struct instance in the store [a weak reference](https://github.com/CharlieTap/chasm/blob/a452ab9e37504517b43118f636a28f54e7bcb57c/executor/runtime/src/commonMain/kotlin/io/github/charlietap/chasm/executor/runtime/store/Store.kt#L24). The idea being that as soon as the execution value is popped from the stack (or is mutated out from a global) then the GC will be able to reclaim the structs/array. 

Is there any guidance around this as I'm not sure if I'm violating an invariant by having some of the stores state reclaimed early? Would be interested to know how others tackled it?

**Comment by titzer** [comment]
A typical runtime won't reified store objects. Rather, the execution stack is generally the source of all roots, and activation frames contain the locals and operand stack. Typically at the implementation level there is an additional reference to an object that represents an instance, and the instance has references to tables and globals. Thus if you consider the execution stack to be the source of all roots and trace through instances, then there is no need to consider some references to be weak.

Often, an optimizing compiler may shorten the lifetimes of locals and even operand slots, so it's possible that even locals do not represent real roots. Since Wasm GC currently has no way to observe collection of objects, it's an unobservable optimization to collect objects early.

**Reply by CharlieTap** [reply]
About this part:

> A typical runtime won't reified store objects.

Do you mean for GC "objects" i.e. struct/array instances? 

If so yeah I guess I could entirely omit struct/arrays instances from the store as ultimately the weak/strong reference setup I currently have achieves the same result (making the execution stack the source of all roots) and only maintains those instances in the store for consistency with the spec. That would save me a lookup also for all struct/array instructions.

For everything else like function instances etc I'm generally following the documention of wasmtimes Store

> A [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) is intended to be a short-lived object in a program. No form of GC is implemented at this time so once an instance is created within a [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) it will not be deallocated until the [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) itself is dropped. This makes [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) unsuitable for creating an unbounded number of instances in it because [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) will never release this memory. It’s recommended to have a [Store](https://docs.wasmtime.dev/api/wasmtime/struct.Store.html) correspond roughly to the lifetime of a “main instance” that an embedding is interested in executing.
